src/java/org/apache/cassandra/index/sai/disk/hnsw/CassandraOnHeapHnsw.java

```diff
@@ -273,10 +275,18 @@ public PriorityQueue<T> search(float[] queryVector, int limit, Bits toAccept, in
         {
             throw new RuntimeException(e);
         }
-        PriorityQueue<T> keyQueue = new PriorityQueue<>();
+        PriorityQueue<PrimaryKey> keyQueue = new PriorityQueue<>();
         while (queue.size() > 0)
         {
-            keyQueue.addAll(keysFromOrdinal(queue.pop()));
+            int node = queue.topNode();
+            float score = queue.topScore();
+            queue.pop();
+            Collection<T> keys = keysFromOrdinal(node);
+            for (T key : keys)
+            {
+                assert key instanceof PrimaryKey;
```

Comment:
Looking at this, I question whether the class needs to be generic anymore. Couldn't we just make it specific to `PrimaryKey`, it is always created with `PrimaryKey`.

Reply:
Equally, if we made this change, there is no reason why this method can't return `PriorityQueue<ScoredPrimaryKey>`. This would avoid the `instanceof` and casting in `QueryContext.recordScores`.

Reply:
We have `CassandraOnHeapHnsw<Integer>`. It is just used to build the graph, not to search it, which is why I could change the return type on this method. Perhaps it'd be worth making `CassandraOnHeapHnsw<T>` abstract with only the relevant shared methods to build the graph, and then the `CassandraOnHeapPrimaryKeyHnsw` implementation would have the search methods.

https://github.com/datastax/cassandra/blob/f0771c5226979392ae6c9c3984f2dd7985ba4775/src/java/org/apache/cassandra/index/sai/disk/v1/SegmentBuilder.java#L173-L200

Reply:
Sorry, I missed that one when I was looking at this. It just seemed to me that we are going through a number of hoops that would be unnecessary if this wasn't generic.